### ISSUE TITLE
[Topbar+] IconController:setTopbarEnabled may not work with duplicate GUI (Explanation on desc.)

### ISSUE THREAD

**Issue by MonsterzSlayerzX65** [1ForeverHD/HDAdmin#42]
My issue is that when I call `IconController:setTopbarEnabled(false)`, the topbar doesn't disappear.

### **Situation and the reason why**

Topbar+ is in `StarterGui`, but while my game is loading, the character doesn't spawn, so to remedy this, I clone the Topbar+ GUI into `PlayerGui`, then I require the module, create my buttons and load everything else. When it finishes, I have my loading script invoke the server to spawn the character, however, when I call `setTopbarEnabled`, it doesn't work. The first thing I did is to check my script for any errors. Then your module's code, I found a `WaitForChild` call in the `IconController` module (function `getTopbarPlusGui`, so I jumped to the conclusion to check my `PlayerGui` and I found that the Topbar+ GUI was duplicated. [The icons were created in one GUI, but `WaitForChild` is returning the other one])

As an improvement and a fix for this issue is to have `IconController` grab the GUI from `Icon` instead of using a `WaitForChild` call,

TLDR / I don't understand: If the GUI is duplicated then there is a chance that `PlayerGui:WaitForChild("Topbar+")` doesn't return the correct one, making `setTopbarEnabled` disable the improper one.

* **Tell me if I should make a pull request for this instead.**
* _**Sorry for the ambiguity and/or bad explanation, English is not my first language.**_

**Comment by Lucke0051** [comment]
I can't replicate. Here is a place where it works for me.
Wait 5 sec and you will see the topbar go away and then 5 sec after it will come back.
https://drive.google.com/file/d/16v4vmxysWvlYb9KI1Ll3w93C3Iy0mZHF/view?usp=sharing

**Comment by Lucke0051** [comment]
Though maybe placing the guis in replicated storage or something, then letting a script on the client clone it would be good. As that would not require the player to spawn either, and it saves server resources.

**Comment by 1ForeverHD** [comment]
This sounds as if the [DirectoryService](https://github.com/1ForeverHD/HDAdmin/tree/master/Utility/DirectoryService) is incorrectly loading Topbar+ due to your particular method, as apposed to there being an issue with the Icon and ``icon:setTopbarEnabled()``.

Are you able to provide a repo place file of this bug?

**Comment by MonsterzSlayerzX65** [comment]
Small video of the issue in action:
**(Had to zip it because otherwise GitHub would not let me upload it as OBS recorded it as a .mkv file)**
[ReproVideo.zip](https://github.com/1ForeverHD/HDAdmin/files/4987092/ReproVideo.zip)

Seems to place-specific as a baseplate doesn't give out this issue.



Test button code:
```
local c = true

script.Parent.MouseButton1Click:Connect(function()
	c = not c
	
	if c then
		script.Parent.Text = "Disable Topbar"
	else
		script.Parent.Text = "Enable Topbar"
	end
	
	local Test = game:GetService("Players").LocalPlayer.PlayerGui:WaitForChild("Topbar+")
	
	if not Test:FindFirstChild("WaitReturn") then
		local x = Instance.new("StringValue")
		x.Name = "WaitReturn"
		x.Parent = Test
	end
	
	print("In")
	local BaseHD = game:GetService("ReplicatedStorage"):WaitForChild("HDAdmin"):WaitForChild("Topbar+")
	local Topbar = require(BaseHD:WaitForChild("IconController"))
	
	Topbar:setTopbarEnabled(c)
	print("Out")
end)
```

**Comment by 1ForeverHD** [comment]
Your client icon creator is fine, it's how you're loading Topbar+ which is almost certainly the issue, in particular this:

![image](https://user-images.githubusercontent.com/51117782/88656679-9af4ea00-d0c8-11ea-8c32-5271363b852a.png)

The Topbar+ server (i.e. DirectoryService) automatically handles the cloning of GUIs into a player's PlayerGUI. By cloning this yourself, the Topbar+ GUI is being cloned twice, which is creating all the issues mentioned above.

For an ideal setup:
1. Topbar+ can be initiated in a Script in ServerScriptService with ``require(4874365424)``
2. Your icon creator local script should be placed within StarterPlayerScripts

**Comment by MonsterzSlayerzX65** [comment]
Ok, reading `DirectoryService`, the conflict is that `DirectoryService` initializes the GUI into `PlayerGui` as you say, however, it **_waits until a character is spawned._** In my loading script, I clone it **BEFORE** the character spawns, and since  `DirectoryService` runs on the server, it will not see the client copy of the GUI, therefore cloning it again.

![RobloxStudioBeta_0EZDZrBfn5](https://user-images.githubusercontent.com/66188098/88663052-35f6bf80-d0db-11ea-9d12-51ef7446a450.png)

**Comment by 1ForeverHD** [comment]
Due to the way Roblox initialises their player, the 'early-joiners' system within the DirectoryService has to wait until the character has loaded plus one heartbeat before cloning in the Topbar+ GUI. 

Can you explain in more detail why you're creating a duplicate in the first place? Can you not simply wait until the original GUI is loaded in?

**Comment by MonsterzSlayerzX65** [comment]
I think it is cleaner to load the topbar now rather than on character spawn, so when the character is spawned, everything is already loaded in, also, I have my Backpack, Admin Panel and Main UI modules create the icons as part of the game loading process (not to mention that my Backpack script hooks the B button to the Icon).

**Comment by 1ForeverHD** [comment]
Without forking from Topbar+ at this point in time to create your desired effect, there isn't much else that can be done. Do consider, Topbar+ only waits for the character for players who join really fast (i.e. maybe the first 1-2 players at most). For others, the Topbar+ GUI is added when their player instance is added.

In the future, my plan is to scrap the DirectoryService completely and have projects (such as Topbar+) initialised in studio via a plugin, as apposed to live in-game through a MainModule, to help in situations like yours.

**Comment by MonsterzSlayerzX65** [comment]
Ok, thanks for the information. Meanwhile, I will use the solution to remove the GUI that DirectoryService clones, as far as I tested, removing the duplicate solves the main problem: `setTopbarEnabled`

Thanks for the information. I will close this issue for now.
